rust/examples/src/llm_dataset_creation.rs

```diff
@@ -0,0 +1,227 @@
+use hf_hub::{api::sync::Api, Repo, RepoType};
+use lance::dataset::WriteParams;
+use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;
+use std::fs::File;
+use std::io::Write;
+use std::sync::Arc;
+use tempfile::NamedTempFile;
+use tokenizers::Tokenizer;
+
+// Implement a custom stream batch reader
+struct WikiTextBatchReader {
+    schema: Arc<Schema>,
+    parquet_readers: Vec<Option<ParquetRecordBatchReaderBuilder<File>>>,
+    current_reader_idx: usize,
+    current_reader: Option<Box<dyn RecordBatchReader + Send>>,
+    tokenizer: Tokenizer,
+    num_samples: u64,
+    cur_samples_cnt: u64,
+}
+
+// Implement Send for WikiTextBatchReader
+// Inform the compiler that type can be transferred between threads
+unsafe impl Send for WikiTextBatchReader {}
+
+impl WikiTextBatchReader {
+    fn new(
+        parquet_readers: Vec<ParquetRecordBatchReaderBuilder<File>>,
+        tokenizer: Tokenizer,
+        num_samples: Option<u64>,
+    ) -> Result<Self> {
+        let schema = Arc::new(Schema::new(vec![Field::new(
+            "input_ids",
+            DataType::List(Arc::new(Field::new("item", DataType::Int64, true))),
+            false,
+        )]));
+
+        Ok(Self {
+            schema,
+            parquet_readers: parquet_readers.into_iter().map(Some).collect(),
+            current_reader_idx: 0,
+            current_reader: None,
+            tokenizer,
+            num_samples: num_samples.unwrap_or(100_000),
+            cur_samples_cnt: 0,
+        })
+    }
+
+    fn process_batch(
+        &mut self,
+        input_batch: &RecordBatch,
+    ) -> Result<RecordBatch, arrow::error::ArrowError> {
+        let num_rows = input_batch.num_rows();
+        let mut token_builder = ListBuilder::new(Int64Builder::with_capacity(num_rows * 1024)); // Pre-allocate space
+        let mut should_break = false;
+
+        if let Some(column) = input_batch.column_by_name("text") {
+            if let Some(string_array) = column.as_any().downcast_ref::<arrow::array::StringArray>()
+            {
+                // Process all rows in the batch
+                for i in 0..num_rows {
+                    if self.cur_samples_cnt >= self.num_samples {
+                        should_break = true;
+                        break;
+                    }
+                    if !Array::is_null(string_array, i) {
+                        let text = string_array.value(i);
+                        if let Ok(encoding) = self.tokenizer.encode(text, true) {
+                            let tb_values = token_builder.values();
+                            for &id in encoding.get_ids() {
+                                tb_values.append_value(id as i64);
+                            }
+                            token_builder.append(true);
+                            self.cur_samples_cnt += 1;
+                            if self.cur_samples_cnt % 1000 == 0 {
+                                println!("Processed {} rows", self.cur_samples_cnt);
```

Comment:
It runs, but much slower than the  python example... I believe I'm doing something wrong here

Reply:
update: release  build is 10x faster. interesting